Mathlib/LinearAlgebra/Matrix/SVD/svdReindex.lean

```diff
@@ -0,0 +1,122 @@
+
+  Note that we know R ≤ MIN(M, N) -/
+variable {𝕂: Type}[IsROrC 𝕂][DecidableEq 𝕂]
+variable {M N: ℕ}
+
+open Matrix BigOperators
+
+namespace Matrix
+
+open scoped ComplexOrder
+
+lemma rank_eq_card_pos_eigs_conj_transpose_mul_self (A: Matrix (Fin M) (Fin N) 𝕂) :
+    A.rank = Fintype.card {i // (isHermitian_transpose_mul_self A).eigenvalues i ≠ 0} := by
+  rw[← rank_conjTranspose_mul_self, IsHermitian.rank_eq_card_non_zero_eigs]
+
+lemma rank_eq_card_pos_eigs_self_mul_conj_transpose (A: Matrix (Fin M) (Fin N) 𝕂) :
+    A.rank = Fintype.card {i // (isHermitian_mul_conjTranspose_self A).eigenvalues i ≠ 0} := by
+  rw[← rank_self_mul_conjTranspose, IsHermitian.rank_eq_card_non_zero_eigs]
+
+/-- For matrix A of size m × n and rank A.rank : we have an equivalence relation between elements of
+Fin (A.rank) and the set non-zero eigenvalues of the matrix Aᴴ⬝A -/
+noncomputable def fin_rank_equiv_eigs_conjTranspose_mul_self (A: Matrix (Fin M) (Fin N) 𝕂) :
+    {i // (isHermitian_transpose_mul_self A).eigenvalues i ≠ 0} ≃ Fin (A.rank) := by
+  apply Fintype.equivFinOfCardEq
+  rw [rank_eq_card_pos_eigs_conj_transpose_mul_self]
+
+/-- For matrix A of size m × n and rank A.rank : we have an equivalence relation between elements of
+Fin (A.rank) and the set non-zero eigenvalues of the matrix A⬝Aᴴ -/
+noncomputable def fin_rank_equiv_eigs_mul_conjTranspose (A: Matrix (Fin M) (Fin N) 𝕂) :
+    {i // (isHermitian_mul_conjTranspose_self A).eigenvalues i ≠ 0} ≃ Fin (A.rank) := by
+  apply Fintype.equivFinOfCardEq
+  rw [rank_eq_card_pos_eigs_self_mul_conj_transpose]
+
+/-- For matrix of size m × n nad rank A.rank : we have an equivalence relation between the elements
+of Fin (width) and the eigevalues of the matrix Aᴴ⬝A partitioned into
+- non-zero eigenvaules: (exactly A.rank) of them see `fin_rank_equiv_eigs_conjTranspose_mul_self`
+- zero eigenvaules: (exactly width - A.rank) of them -/
+noncomputable def equiv_fin_width_eigs_conjTranspose_mul_self (A: Matrix (Fin M) (Fin N) 𝕂) :
+    (Fin N) ≃ (Fin A.rank) ⊕ (Fin (N - A.rank)) := by
+  let em := Equiv.sumCompl (fun i =>  (isHermitian_transpose_mul_self A).eigenvalues i ≠ 0)
+  let eₙᵣ : {i // ¬(isHermitian_transpose_mul_self A).eigenvalues i ≠ 0} ≃ Fin (N - A.rank) := by
+    apply Fintype.equivFinOfCardEq
+    rw [Fintype.card_subtype_compl, Fintype.card_fin, rank_eq_card_pos_eigs_conj_transpose_mul_self]
+  exact Equiv.trans em.symm  (Equiv.sumCongr (fin_rank_equiv_eigs_conjTranspose_mul_self A) eₙᵣ)
+
+/-- For matrix of size m × n nad rank A.rank : we have an equivalence relation between the elements
+of Fin (height) and the eigevalues of the matrix A⬝Aᴴ partitioned into
+- non-zero eigenvaules: (exactly A.rank) of them see `fin_rank_equiv_eigs_mul_conjTranspose`
+- zero eigenvaules: (exactly height - A.rank) of them -/
+noncomputable def equiv_fin_height_eigs_mul_conjTranspose (A: Matrix (Fin M) (Fin N) 𝕂) :
+    (Fin M) ≃ (Fin A.rank) ⊕ (Fin (M - A.rank)) := by
```

Comment:
This looks like a special case of [`finSumFinEquiv`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Fin.html#finSumFinEquiv) to me, using `Fin.cast` to first turn `Fin M` into `Fin (A.rank + (M - A.rank))`

Reply:
Just for other people who pass by.!
Contrary to what eric says here, this is not a special case o `finSumFinEquiv`.  The permutation splitting `Fin (M)` is not any permutation. It must split the zero eigenvalues and and non-zero  eigenvalues alone. Instead of doing all the work in `{ i // eig \neq 0 }` and `{ i // eig = 0 }` we recast back to `Fin (A.rank)` and `Fin (M - A.rank)` which are a nicer interface for the API of the SVD and it matches more closely what people do on paper when proving SVD or using it in another proof.

Reply:
I agree with Mohanad, the type might be the same but the value is different. `finSumFinEquiv` would send the first `r` values to the left part of the sum and the rest to the right. This is a more specific operation.